File: apple/MeshDropMac/Sources/mock/EngineProjection.swift
import Foundation
import SwiftUI
import MeshDropKit

// MARK: - Engine 类型 → UI mock-shape 类型 的投影
//
// UI 阶段是直接用 MockDevice / MockHistory / MockPendingPairing / MockPendingOffer /
// MockTrustedDevice 这些视觉密集的结构体绘制的（带颜色、initials、角度等）。
// backend 接入轮不重做视觉，转而把 ShareEngine 的真实类型投影到这些 mock 视觉类型上，
// 让原 UI 代码改最少的引用即可。
//
// Mock 结构体本身保留给 SwiftUI Preview 用。

extension MockDevice {
    /// 把 Engine 的 Device 投影成 UI 用的 MockDevice。
    /// 显示用的 angle / dist / color / initials 由 id 哈希稳定派生。
    static func from(_ device: Device, online: Bool = true) -> MockDevice {
        let initials = makeInitials(device.name)
        let kind = mapKind(device.os)
        let osLabel = osLabel(device.os, model: device.model)
        let stableAngle = stableDouble(device.id, salt: "angle", in: 0..<360)
        let stableDist = stableDouble(device.id, salt: "dist", in: 0.3..<0.92)
        let color = stableColor(device.id)
        return MockDevice(
            id: device.id,
            name: device.name,
            who: device.name,
            kind: kind,
            dist: stableDist,
            angle: stableAngle,
            color: color,
            initials: initials,
            os: osLabel,
            rtt: stableRtt(device.id),
            online: online
        )
    }
}

extension MockHistory {
    /// 把 Engine 的 HistoryItem 投影成 UI 用的 MockHistory。
    static func from(_ item: HistoryItem) -> MockHistory {
        let dir: HistoryDir = (item.direction == .outgoing) ? .outgoing : .incoming
        let time = Self.timeFormatter.string(from: item.createdAt)
        let status: HistoryStatus
        switch item.status {
        case .completed: status = .done
        case .transferring: status = .transferring
        case .pending, .waitingApproval: status = .queued
        case .failed, .canceled: status = .failed
        }
        switch item.kind {
        case .text(let s):
            return MockHistory(
                id: item.id.uuidString,
                dir: dir,
                peer: item.peer.name,
                time: time,
                kind: .text,
                content: s,
                status: status
            )
        case .file(let name, let size, _):
            let ext = (name as NSString).pathExtension.lowercased()
            let sizeStr = byteFormatter.string(fromByteCount: Int64(size))
            let progress: Int? = {
                if case let .transferring(done, total) = item.status, total > 0 {
                    return Int(Double(done) / Double(total) * 100)
                }
                return nil
            }()
            return MockHistory(
                id: item.id.uuidString,
                dir: dir,
                peer: item.peer.name,
                time: time,
                kind: .file,
                name: name,
                size: sizeStr,
                ext: ext.isEmpty ? "bin" : ext,
                progress: progress,
                status: status
            )
        }
    }

    private static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm"
        return f
    }()
}

extension MockPendingPairing {
    static func from(_ req: PairingRequest) -> MockPendingPairing {
        MockPendingPairing(
            id: req.id.uuidString,
            peer: req.peer.name,
            deviceName: req.peer.model ?? req.peer.name,
            fingerprint: req.peer.humanFingerprint,
            receivedAt: shortAgo(from: req.receivedAt)
        )
    }
}

extension MockPendingOffer {
    static func from(_ offer: PendingFileOffer) -> MockPendingOffer {
        MockPendingOffer(
            id: offer.id.uuidString,
            peer: offer.peer.name,
            deviceName: offer.peer.model ?? offer.peer.name,
            fileName: offer.fileName,
            fileSize: offer.formattedSize,
            note: "",
            receivedAt: shortAgo(from: offer.receivedAt)
        )
    }
}

extension MockTrustedDevice {
    static func from(_ record: TrustRecord, online: Bool = false, kind: DeviceKind = .mac) -> MockTrustedDevice {
        MockTrustedDevice(
            id: record.fingerprint,
            name: record.name,
            who: record.name,
            kind: kind,
            fingerprint: humanize(record.fingerprint),
            pairedAt: dateFormatter.string(from: record.firstSeen),
            lastSeen: shortAgo(from: record.lastSeen),
            online: online
        )
    }

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()
}

// MARK: - 辅助

private let byteFormatter: ByteCountFormatter = {
    let f = ByteCountFormatter()
    f.allowedUnits = [.useKB, .useMB, .useGB]
    f.countStyle = .file
    return f
}()

private func makeInitials(_ name: String) -> String {
    let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.isEmpty { return "?" }
    let parts = trimmed.split(separator: " ")
    if parts.count >= 2 {
        let a = parts[0].first.map { String($0) } ?? ""
        let b = parts[1].first.map { String($0) } ?? ""
        return (a + b).uppercased()
    }
    // 中文 / 单 token：取头两字
    return String(trimmed.prefix(2))
}

private func mapKind(_ os: DeviceOS) -> DeviceKind {
    switch os {
    case .ios: return .ios
    case .android: return .android
    case .macos: return .mac
    case .windows: return .win
    case .linux: return .mac
    }
}

private func osLabel(_ os: DeviceOS, model: String?) -> String {
    if let model, !model.isEmpty { return model }
    switch os {
    case .ios: return "iOS"
    case .android: return "Android"
    case .macos: return "macOS"
    case .windows: return "Windows"
    case .linux: return "Linux"
    }
}

private func stableDouble(_ id: String, salt: String, in range: Range<Double>) -> Double {
    let seed = abs((id + "|" + salt).hashValue)
    let span = range.upperBound - range.lowerBound
    return range.lowerBound + Double(seed % 1000) / 1000.0 * span
}

private let kPalette: [UInt32] = [
    0xFFB4A1, 0xB7E5C8, 0xC7B8FF, 0xFFD970, 0x9AD0FF,
    0xFFC4DC, 0xCFE8B0, 0xFFD1A6, 0xA9E0E0
]

private func stableColor(_ id: String) -> Color {
    let i = abs(id.hashValue) % kPalette.count
    return Color(hex: kPalette[i])
}

private func stableRtt(_ id: String) -> Int {
    Int(stableDouble(id, salt: "rtt", in: 8..<48))
}

private func shortAgo(from date: Date) -> String {
    let s = max(0, Int(Date().timeIntervalSince(date)))
    if s < 60 { return "\(s)s 前" }
    let m = s / 60
    if m < 60 { return "\(m) 分钟前" }
    let h = m / 60
    if h < 24 { return "\(h) 小时前" }
    return "\(h / 24) 天前"
}

private func humanize(_ fp: String) -> String {
    let upper = fp.uppercased()
    var out: [String] = []
    var idx = upper.startIndex
    while idx < upper.endIndex {
        let end = upper.index(idx, offsetBy: 4, limitedBy: upper.endIndex) ?? upper.endIndex
        out.append(String(upper[idx..<end]))
        idx = end
    }
    return out.joined(separator: " · ")
}

File: windows/MeshDrop/ViewModels/EngineProjection.cs
using System;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Linq;
using MeshDrop.Mock;
using MeshDrop.Models;
using MeshDrop.Transport;

namespace MeshDrop.ViewModels;

/// <summary>
/// 把 ShareEngine 的真实模型 (Device / HistoryItem) 投影成 view-DTO
/// (MockDevice / MockHistory / MockTransfer / MockMessage / MockTrust)。
///
/// 数据源真实来自 engine — view-DTO 类型只是 XAML / Controls 复用的现成 shape。
/// MockData.cs 静态字段不再被 runtime 引用，仅留给 SwiftUI/WinUI Preview。
/// </summary>
internal static class EngineProjection
{
    // ─── Device → MockDevice ──────────────────────────────────────

    public static MockDevice ToMock(this Device d)
    {
        var kind = d.Os switch
        {
            DeviceOS.Macos => MockKind.Mac,
            DeviceOS.Windows => MockKind.Win,
            DeviceOS.Ios => MockKind.Ios,
            DeviceOS.Android => MockKind.Android,
            DeviceOS.Linux => MockKind.Linux,
            _ => MockKind.Linux,
        };
        var initials = Initials(d.Name);
        var color = ColorFromId(d.Id);
        var (angle, dist) = RadarSlot(d.Id);
        var osLabel = d.Os switch
        {
            DeviceOS.Macos => "macOS",
            DeviceOS.Windows => "Windows",
            DeviceOS.Ios => "iOS",
            DeviceOS.Android => "Android",
            DeviceOS.Linux => "Linux",
            _ => "Linux",
        };
        return new MockDevice(
            Id: d.Id,
            Name: d.Name,
            Who: d.Name,
            Kind: kind,
            Dist: dist,
            Angle: angle,
            ColorHex: color,
            Initials: initials,
            Os: d.Model is null ? osLabel : $"{osLabel} · {d.Model}",
            Rtt: 0,
            Fingerprint: d.HumanFingerprint,
            Ip: d.Host ?? "—",
            Bw: "LAN",
            E2EVerified: true);
    }

    public static string Initials(string name)
    {
        var n = (name ?? "").Trim();
        if (n.Length == 0) return "·";
        var parts = n.Split(new[] { ' ', '·', '-', '_' }, StringSplitOptions.RemoveEmptyEntries);
        if (parts.Length == 0) return n.Substring(0, Math.Min(2, n.Length)).ToUpperInvariant();
        if (parts.Length == 1) return parts[0].Substring(0, Math.Min(2, parts[0].Length)).ToUpperInvariant();
        return (parts[0][0].ToString() + parts[1][0]).ToUpperInvariant();
    }

    public static string ColorFromId(string id)
    {
        string[] palette =
        {
            "#FFB4A1", "#B7E5C8", "#C7B8FF", "#FFD970", "#9AD0FF",
            "#FF9EC4", "#A0E0D6", "#FFC68A", "#C2D6FF", "#EBC0FF",
        };
        var h = 0;
        foreach (var c in id) h = (h * 31 + c) & 0x7FFFFFFF;
        return palette[h % palette.Length];
    }

    private static (int angle, double dist) RadarSlot(string id)
    {
        var h = 0;
        foreach (var c in id) h = (h * 131 + c) & 0x7FFFFFFF;
        var angle = h % 360;
        var dist = 0.32 + (h / 360 % 100) / 100.0 * 0.55;
        return (angle, dist);
    }

    // ─── HistoryItem → MockHistory ────────────────────────────────

    public static MockHistory ToMock(this HistoryItem h)
    {
        var dir = h.Direction == TransferDirection.Outgoing ? "outgoing" : "incoming";
        switch (h.Kind)
        {
            case HistoryKind.Text t:
                return new MockHistory(
                    Id: h.Id.ToString(), Dir: dir, Peer: h.Peer.Name,
                    Time: h.FormattedTime, Kind: MockHistoryKind.Text,
                    Content: t.Content, Status: StatusStr(h.Status), Progress: ProgressPct(h.Status));
            case HistoryKind.File f:
                var ext = System.IO.Path.GetExtension(f.Name).TrimStart('.');
                return new MockHistory(
                    Id: h.Id.ToString(), Dir: dir, Peer: h.Peer.Name,
                    Time: h.FormattedTime, Kind: MockHistoryKind.File,
                    Name: f.Name, Size: ByteFormat.Format(f.Size),
                    Ext: ext, Status: StatusStr(h.Status), Progress: ProgressPct(h.Status));
            default:
                return new MockHistory(
                    Id: h.Id.ToString(), Dir: dir, Peer: h.Peer.Name,
                    Time: h.FormattedTime, Kind: MockHistoryKind.Text);
        }
    }

    public static MockTransfer ToTransfer(this HistoryItem h)
    {
        var name = h.Kind switch
        {
            HistoryKind.File f => f.Name,
            HistoryKind.Text t => Truncate(t.Content, 24),
            _ => "(unknown)",
        };
        var size = h.Kind switch
        {
            HistoryKind.File f => ByteFormat.Format(f.Size),
            HistoryKind.Text t => ByteFormat.Format(System.Text.Encoding.UTF8.GetByteCount(t.Content)),
            _ => "0 B",
        };
        var ext = h.Kind switch
        {
            HistoryKind.File f => System.IO.Path.GetExtension(f.Name).TrimStart('.'),
            _ => "txt",
        };
        var me = MeshDrop.I18n.T("common.me");
        var from = h.Direction == TransferDirection.Outgoing ? me : h.Peer.Name;
        var to = h.Direction == TransferDirection.Outgoing ? h.Peer.Name : me;
        var (state, progress, speed) = h.Status switch
        {
            TransferStatus.Pending => (MockTransferState.Queued, 0, (string?)null),
            TransferStatus.WaitingApproval => (MockTransferState.Queued, 0, (string?)null),
            TransferStatus.Transferring tr => (
                h.Direction == TransferDirection.Outgoing ? MockTransferState.Sending : MockTransferState.Receiving,
                (int)(tr.Fraction * 100), (string?)"—"),
            TransferStatus.Completed => (MockTransferState.Done, 100, (string?)null),
            TransferStatus.Failed => (MockTransferState.Failed, 0, (string?)null),
            TransferStatus.Canceled => (MockTransferState.Failed, 0, (string?)null),
            _ => (MockTransferState.Queued, 0, (string?)null),
        };
        return new MockTransfer(name, size, ext, from, to, progress, state, speed, Id: h.Id.ToString());
    }

    public static MockMessage ToMessage(this HistoryItem h)
    {
        var side = h.Direction == TransferDirection.Outgoing ? "out" : "in";
        var delivered = h.Status is TransferStatus.Completed;
        var time = h.CreatedAt.ToString("HH:mm");
        switch (h.Kind)
        {
            case HistoryKind.Text t:
                return new MockMessage(h.Id.ToString(), side, "text",
                    Text: t.Content, Time: time, Delivered: delivered);
            case HistoryKind.File f:
                return new MockMessage(h.Id.ToString(), side, "file",
                    FileName: f.Name, FileSize: ByteFormat.Format(f.Size),
                    FileExt: System.IO.Path.GetExtension(f.Name).TrimStart('.'),
                    Time: time, Delivered: delivered);
            default:
                return new MockMessage(h.Id.ToString(), side, "text", Time: time, Delivered: delivered);
        }
    }

    // ─── ClipboardEntry → MockClipboard ───────────────────────────

    public static MockClipboard ToMock(this ClipboardEntry e)
    {
        var kind = e.Kind switch
        {
            "link" => MockClipboardKind.Link,
            "code" => MockClipboardKind.Code,
            _ => MockClipboardKind.Text,   // 未知 / "text" 都归 Text
        };
        return new MockClipboard(
            Id: e.Id.ToString(),
            Who: e.PeerName,
            Kind: kind,
            Body: e.Content ?? "",
            Ago: e.ReceivedAt.ToString("HH:mm:ss"));
    }

    public static MockTrust ToMock(this TrustRecord t)
    {
        var date = DateTimeOffset.FromUnixTimeMilliseconds(t.LastSeenMs).LocalDateTime;
        return new MockTrust(
            Who: t.Name, DeviceName: t.Name, Os: "",
            Fingerprint: HumanFingerprint(t.Fingerprint),
            PairedOn: date.ToString("yyyy-MM-dd"),
            LastSeen: date.ToString("HH:mm"),
            RememberAllowed: true);
    }

    private static string HumanFingerprint(string fp)
    {
        var up = (fp ?? "").ToUpperInvariant();
        var sb = new System.Text.StringBuilder();
        for (var i = 0; i < up.Length; i += 4)
        {
            if (i > 0) sb.Append(" · ");
            sb.Append(up.Substring(i, Math.Min(4, up.Length - i)));
        }
        return sb.ToString();
    }

    private static string StatusStr(TransferStatus s) => s switch
    {
        TransferStatus.Pending => "queued",
        TransferStatus.WaitingApproval => "queued",
        TransferStatus.Transferring => "transferring",
        TransferStatus.Completed => "done",
        TransferStatus.Failed => "failed",
        TransferStatus.Canceled => "failed",
        _ => "queued",
    };

    private static int? ProgressPct(TransferStatus s) =>
        s is TransferStatus.Transferring tr ? (int)(tr.Fraction * 100) : null;

    private static string Truncate(string s, int n) =>
        string.IsNullOrEmpty(s) ? "" : (s.Length <= n ? s : s.Substring(0, n) + "…");
}

/// <summary>
/// 把 ShareEngine 的 ObservableCollection&lt;TSource&gt; 实时投影成
/// ObservableCollection&lt;TView&gt;，XAML 直接绑定 view 集合。
/// </summary>
internal sealed class ProjectedCollection<TSource, TView> : ObservableCollection<TView>
    where TSource : class
{
    private readonly Func<TSource, TView> _map;
    private readonly ObservableCollection<TSource> _src;
    private readonly Func<TSource, bool>? _filter;

    public ProjectedCollection(ObservableCollection<TSource> source,
                               Func<TSource, TView> map,
                               Func<TSource, bool>? filter = null)
    {
        _src = source;
        _map = map;
        _filter = filter;
        Rebuild();
        _src.CollectionChanged += OnSrcChanged;
    }

    private void OnSrcChanged(object? sender, NotifyCollectionChangedEventArgs e)
    {
        // 投影做最简：变更时全量重建。Devices/History 量都不大。
        Rebuild();
    }

    public void Refresh() => Rebuild();

    private void Rebuild()
    {
        Clear();
        foreach (var s in _src)
            if (_filter is null || _filter(s))
                Add(_map(s));
    }
}
